homeassistant/components/madvr/coordinator.py

```diff
@@ -0,0 +1,66 @@
+"""Coordinator for handling data fetching and updates."""
+
+import logging
+from typing import TYPE_CHECKING, Any
+
+from madvr.madvr import Madvr
+
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_PORT
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+if TYPE_CHECKING:
+    from . import MadVRConfigEntry
+
+
+class MadVRCoordinator(DataUpdateCoordinator[dict[str, Any]]):
+    """Madvr coordinator for Envy (push-based API)."""
+
+    if TYPE_CHECKING:
+        config_entry: MadVRConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: Madvr,
+    ) -> None:
+        """Initialize madvr coordinator."""
+        super().__init__(hass, _LOGGER, name=DOMAIN)
+        self.entry_id = self.config_entry.entry_id
+        # get the mac address from the config entry
+        self.mac = self.config_entry.data.get(CONF_MAC, "")
+        self.client = client
+        self.client.set_update_callback(self.handle_push_data)
+        _LOGGER.debug("MadVRCoordinator initialized with mac: %s", self.mac)
+
+    @property
+    def device_info(self) -> DeviceInfo:
```

Comment:
I think we can move this into a shared base entity in the future, for now you can also just keep it in the remote